Vet clinic database: database performance audit

mwismann opened the pull request:
# Vet clinic database: database performance audit
**The following changes were implemented:**

### On `README.md`:

- Added information about second author ✅ 
- Modified `Future Features` ✅ 

### On `LICENSE`:

- Added second author's name ✅ 

### On `schema.sql`:

- Added an `email` column to the `owners` table ✅ 
- Created an `index table` for the `visits` table's `animal_id` ✅ 
- Created an `index table` for the `visits` table ✅ 

### On `data.sql`:

- Inserted around **21,565,728** rows into the `visits` table ✅ 
- Inserted around **15,000,000** rows into the `owners` table ✅ 
- _**THE FOLLOWING SCREENSHOT DISPLAYS THE NEED FOR SEVERAL INSERTIONS OF DATA IN ORDER FOR US 

TO RUN AN `EXPLAIN ANALYZE` THAT COULD RETURN AN `EXECUTION TIME X >= 1000 MS`**_ 🤠 
![insert_data](https://github.com/mwismann/vet_clinic_2/assets/107666488/07388b26-1701-4ca0-816f-15ab9c972975)


### On `queries.sql`:

- **Ran a performance audit for the following queries:**

1. `EXPLAIN ANALYZE SELECT COUNT(*) FROM visits where animal_id = 4;`

- BEFORE: `*Execution Time X > 1300 ms*`
![q1_before](https://github.com/mwismann/vet_clinic_2/assets/107666488/257805fd-d36a-40b1-82da-a3d6d0acbf4a)

- AFTER: *`Execution Time X < 150 ms`* ✅
![q1_after](https://github.com/mwismann/vet_clinic_2/assets/107666488/36f6ce10-375a-4863-b0b9-21f100d591f9)


2. `EXPLAIN ANALYZE SELECT * FROM visits where vet_id = 2;`

- BEFORE: `*Execution Time X > 5600 ms*`
![q2_before](https://github.com/mwismann/vet_clinic_2/assets/107666488/dcc014f6-dd9e-4b62-b152-e4de202146f4)

- AFTER: *`Execution Time X < 2000 ms`* ✅ 
![q2_after](https://github.com/mwismann/vet_clinic_2/assets/107666488/1be07248-c966-4bab-9b93-7eda0b679a73)

3. `EXPLAIN ANALYZE SELECT * FROM owners where email = 'owner_18327@mail.com';`

- BEFORE: `*Execution Time X > 1400 ms*`
![q3_before](https://github.com/mwismann/vet_clinic_2/assets/107666488/6339df39-8cb8-4a2f-a512-d313dfbd84cd)

- AFTER: *`Execution Time X < 1 ms`* ✅ 
![q3_after](https://github.com/mwismann/vet_clinic_2/assets/107666488/63059283-c956-4616-8991-a382f5ef0da8)

mwismann committed:
Insert data for performance audit

ITurres committed:
feat: on 'schema.sql' added new Indexes to tables

- Added Index 'ix_visit_animal' to table -visit-.
- Added Index 'ix_visits' to table -visit- for multiple columns.
- Added Index 'ix_owner_email' to table -owners-.

ITurres committed:
feat: on 'queries.sql' Added explained analyze to the queries

- Added for table -visits- and -owners- the explain analyze

mwismann committed:
Update README

mwismann committed:
Update LICENSE

Olamarx commented on schema.sql:
```diff
@@ -63,4 +63,13 @@ CREATE TABLE visits (
-)
\ No newline at end of file
+)
+
+-- Add an email column to your owners table
+ALTER TABLE owners ADD COLUMN email VARCHAR(120);
+
+CREATE INDEX ix_visit_animal ON visits (animal_id);
+
+CREATE INDEX ix_visits ON visits (vet_id, id, animal_id, date_of_visit);
+
+CREATE INDEX ix_visits ON visits (vet_id, id, animal_id, date_of_visit);
```
- [x] Thank you for adding the screenshots. I can see that the query planner used index for SELECT * FROM owners where email = 'owner_18327@mail.com' and the time decreased.

But for the other two - nothing has changed.

Can you please run again the index creation queries:

```
CREATE INDEX animal_idx ON visits (animal_id);
CREATE INDEX vet_idx ON visits (vet_id);
```

ITurres replied:
Hi @Olamarx ! thank you for reviewing this pr! 🎯 and for pointing this issue out!

My bad 😄 , I have re coded the same Index creator instead of the correct one.

Changes will be made to mitigate it.

Cheers! 

Done ✅

Olamarx reviewed (changes requested):
Hi Team, 

Good job so far, indeed you have spent time on this project!
There are some issues that you still need to work on to go to the next project but you are almost there!
<img src="https://media3.giphy.com/media/xUOxfh6ZM75efM3Bqo/giphy-downsized-medium.gif"/>

## Highlights
- Readme file professional ✔️

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

ITurres committed:
fix: on 'schema.sql' address code repetition:

- removed the 'ix_visits' repeated Index creator.
- Added the missing 'ix_owner_email' Index creator.

LionRouge1 reviewed (approved):
Hi **Team**, 

# STATUS: APPROVED! 🥇 💯 
As I can clearly see that you had some fun with the `explain analyze` and I am sure that in the future you will know how to start the investigation for some slow DB queries, it's time to approve this project :smiley: 
Your project is complete! There is nothing else to say other than... it's time to merge it :shipit:

![done](https://media.giphy.com/media/a5llhdWKMOIfneMCpB/giphy.gif)

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

ITurres commented:
Hi @LionRouge1 Lion!, together with @mwismann  we want to thank you for reviewing this pull request 🥳! Cheers and happy coding! 👨🏼‍💻

mukhammadrizooff commented on queries.sql:
```diff
@@ -233,3 +233,9 @@ JOIN species ON species.id = animals.species_id
+
+EXPLAIN ANALYZE SELECT COUNT(*) FROM visits WHERE animal_id = 4;
+
+EXPLAIN ANALYZE SELECT * FROM visits where vet_id = 2;
```
- [ ] [REQUIRED] Good job so far! The second screenshot should show how much time the query takes `AFTER` the improvements are introduced by you
- The [execution time should be significantly lower](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_performance_audit.md) than `1 sec (1000 ms)` . Please after implementation this request change the `second` screenshot result 🙏🏼 

![image](https://github.com/mwismann/vet_clinic_2/assets/63915024/abad0d98-9cb7-4a5f-bcfd-e14970ba6c52)
![image](https://github.com/mwismann/vet_clinic_2/assets/63915024/4259be42-5eda-4528-903f-7c9d23ed11aa)

ITurres replied:
Hi, @mukhammadrizooff! I hope you are doing well 🤠

Regarding the project's requirements found at [this link](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_performance_audit.md), the "execution time: X ms and X < 1000" metric pertains to the time it takes to analyze a query after inserting a specific amount of data. **_It's important to note that this measurement doesn't represent the expected execution time after improvements_**. Instead, it should be greater than 1000 ms when first inserting data before any enhancements.

If you take a look at our screenshot, you'll see that after inserting data into the database multiple times, we observed an execution time of 5600 ms, which met the initial requirement for proceeding with the project.

Once the improvements were implemented, we reanalyzed the query and achieved an outstanding execution time of 1968 ms, a significant optimization compared to the initial execution time.

mukhammadrizooff reviewed (commented):
Hi @TEAM, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

<img src="https://media0.giphy.com/media/Jq824R93JsLwZCaiSL/giphy.gif"/>

## Highlights
- No linter errors ✔️ 
- Correct GitFlow ✔️ 
- Good implementations ✔️
- Descriptive `README` file ✔️

## Required Changes ♻️
- [ ] [REQUIRED] Good job so far! Please you should create a new `Pull Request` for the next submission and your implementation should be provided for the next `Code Reviewer` and [`Pick the link`](https://github.com/mwismann/vet_clinic_2/pull/5#pullrequestreview-1704598872) of this review on the new `PR` description 🙏🏼 

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

nicupop729 reviewed (commented):
Hello @mwismann  👋,

_This is an additional review as a result of the second opinion request._

Please be aware that using a previously approved PR is not permitted for pair-programming activities.

To proceed with this milestone, kindly create a new branch with your new pair-programming partner and collaborate on implementing the project requirements together.

The review attempts will be manually reset later today and the last review of @mukhammadrizooff will not be considered.

We apologise for the confusion and thank you for your understanding!

Keep rocking 🚀!

nicupop729 reviewed (commented):
@ITurres and @gdvl777, please note that is not allowed to use other students GitHub repository to ask for review.

In this case, @mwismann owns this GitHub account.
Kindly use one of your own repositories to work on this milestone.

@mukhammadrizooff please note here multiple reasons for invalidation.

ITurres commented:
Hi, @nicupop729 and @mukhammadrizooff ! totally right, my bad! I totally forgot either me or my new cp will need to open a new branch on one of our repos. Sorry again!
